### ISSUE TITLE
directive wiring in code - first

### ISSUE THREAD

**Issue by yarinvak** [graphql-java/graphql-java#1076]
Hey,
Is directive wiring supported in the code first schema creation, and if so - how can I use it (didn't find it on docs)?

Thanks

**Comment by bbakerman** [comment]
No - it does not make sense. 

Directive wiring allows you to gather say "20 fields" and wire them all in one go.  But the API requires you to build 20 fields by hand so you can just put the data fetchers (and hence the behaviour) at the time you build the field.

**Comment by yarinvak** [comment]
I see a little problem that maybe you can give me some solution for.
Lets say I want to build a GraphQLType for the Query type. I can supply this type with 'GraphQLDirective's. I guess these are the directives that are decorating the type itself, like as if I build a GraphQLField with directives.
But the directives definitions should be supplied to the GraphQLSchema, so if the schema does not contain a directive definition, it should not work on the Query type or on some field.
But while building the Query type, or any graphql type, I do not know which directives are declared in the schema. So i might be using directives on some type or field which are not declared in the schema and it will still work.

To be more detailed: in graphql-java-annotations we allow building graphqltypes out of annotated classes and fields. And if i wish to add some directive wiring functionality, i will never know what directives are declared in the schema because its not in my scope. 

What do you think is the right usage in this case?

**Comment by bbakerman** [comment]
I think you would be better to invent a post processing step in graphql-java-annotations much like what we do in the SDL directive wiring.

To explain I will tell you how we build SDL

* Step 1 - parse the SDL in type definitions (and validate as much as we can)
* Step 2a - create the GraphqlType tree from those definitions including default / wired data fetchers
* Step 2b - after the field and containing GraphqlType is built, call out to the DirectiveWiring to mutate it into any form it likes (this is a kinda of post processing
* Step 3 - Complete Schema is now built

You seem to be asking for defaulting mechanism (field -> object type  -> schema) looking for directives / metadata.  This is going to be hard before you have built the complete schema but then circular trees are always tricky like that.

I still feel like the interface `graphql.schema.idl.SchemaDirectiveWiring` is not the right  way for graphql-java-annotations to tweak the built tree.  The directives used are in SDL `graphql.language.Directive` form when used in the schema generator

You could for example synthesise runtime 'graphql.schema.GraphQLDirective` objects when running the graphql-java-annotations processing and then later walk the tree and change the behaviours based on them say.  I am guessing you are transferring "java annotations" into graphql directives so you can store metadata about a type??

Can you give me a more concrete example of Java class -> Graphql type and how you want to process them?

**Comment by yarinvak** [comment]
For example, lets say I have the type Person.
In graphql-java-annotations, I would create a class Person:
```java
public class Person{
@GraphQLField
String name;
@GraphQLField
Person[] friends;
}
```
It will be generated into a corresponding GraphQLType of Person, when I'll call GraphQLAnnotations.object(Person.class).

Let's say I wish to add GraphQLDirective support over one of the Person's fields.
So I will annotate it like that:
```java
@GraphQLField
@GraphQLDirective("toUpperCase")
String name;
```
My dilemma is where to declare the "toUpperCase" directive? The problem is not how to generate a GraphQLDirective from annotation, but where to declare this annotation. Because in SDL we declare the directive definition on the schema itself, but here I have only the Person, and I don't think the Person class is the right place to put the directive definition on.

I guess that any solution is not going to be very clean, because I do not have the schema while creating the type.
I would suggest that GraphQL-Java will validate that if a type or a field in the schema contains directives, they are also declared/defined on the schema, so a situation where field is decorated with an unsupported directive would not be happening.

**Comment by bbakerman** [comment]
> My dilemma is where to declare the "toUpperCase" directive? The problem is not how to generate a GraphQLDirective from annotation, but where to declare this annotation. 

Just create a GraphQLDirective instance and put it on the type at type creation time.  Thats not the hard part.  The hard part is later adding behaviour or mutating types.

> I guess that any solution is not going to be very clean, because I do not have the schema while creating the type.

I think you will want to do a "post schema" creation step where you take all the directives declared and run them through a process where you mix in behaviors.  For example your `toUpperCase` directive will need to change the data fetcher if the field its on.

This post step is exactly what happens in the SDL creation steps.  We "mutate" the types and fields after they are built with the directive wiring and that can add/remove fields and add / remove data fetcher behaviour

**Comment by bbakerman** [comment]
> But directives definitions should be declared on the schema, as far as I understand, and not over specific fields/types.

True but outside of SDL they just don't make as much sense.  I personally would not try to invent a holder of valid places for directives in a runtime only type system.

> So if some field or type is decorated with a directive which is not defined on the root level of the schema which contains the object, an error shoud occur (I think this is a reasonable behaviour).

I would not do that but that is up to you.  You will need to invent a new "valid directive locations holder sub system" because I don't think we would put such a thing into graphql-java itself.  We MAY but my initial inclination is to not add it.

For a start its NOT in the reference implementation (which we take inspiration from although we go beyond it at times) and it just doesn't seem add value to people who create a runtime type system by hand.

I can see how with your @annotation based system you need more constraints on directives but I am not sure it should go into the base graphql-java library

Imagine you had this in the graphql-java-annotations

```
public @interface DirectiveLocation {

    String[] locations();

    String name();
}

public @interface DirectiveLocations {
    DirectiveLocation[] value();
}


    @DirectiveLocations({
            @DirectiveLocation(name = "objectAndFieldDirective", locations = {"OBJECT", "FIELD_DEFINITION"}),
            @DirectiveLocation(name = "objectOnlyDirective", locations = {"OBJECT"}),
    })
    public static class SomeRegistryClass {

    }
}
```

This would allow people to indicate what directive names are allowed where.

Then later in your annotation processing, you would discover any `@DirectiveLocations` and create a list of all the value places.  You would then check them against the built graphql runtime objects BEFORE you build out the schema (this is the post processing step I mention above)

If the locations are valid, you would allow the schema to be built otherwise throw exceptions.

ps.  You can use enums etc.. instead of strings.  I just did that to type less.

@andimarek - what are your thoughts on the idea of having a directive definition constructs attached to the runtime schema?

**Comment by yarinvak** [comment]
I see what you did here and I did implement something like that for the meantime.
But let's talk about graphql-java.

Using your library, someone might create a project which only creates some GraphQLTypes. Another project will create different GraphQLTypes. Than, in a third project, we would merge these types into a schema. 
Now, if in the first project we decorated the types with a directive named "uppercase" with some valid locations/valid arguments, and in the second project we decorated with a directive named "uppercase" too but with other valid locations, we created a conflict. There is not a single point of truth (which in my opinion should be the directives which are declared on the schema itself).
And more than that, in the current implementation we can add directives to fields and types without adding them to the schema, and they won't be exposed on the introspection query.

A validation on the schema.build() which validates that every directive which is being used in a type/field/etc is also declared on the schema, and that there are no multiple directives with the same name, may solve that.
